src/models/Calendar.php

```diff
@@ -0,0 +1,321 @@
+                        default => 0,
+                    },
+                    CalendarKeys::Categories->value,
+                    CalendarKeys::Items->value => is_array($reqBody[$key]) ? array_filter($reqBody[$key], 'is_int') : null,
+                    default => null,
+                };
+            }
+        }
+
+        return $this->create($normalizedReqBody);
+    }
+
+    public function patch(Action $action, array $params): array
+    {
+        throw new ImproperActionException('No PATCH action.');
+    }
+
+    public function getApiPath(): string
+    {
+        return 'api/v2/calendar/';
+    }
+
+    public function readAll(): array
+    {
+        $page = explode('/', (Request::createFromGlobals())->getScriptName());
+        $isProfile = end($page) === 'profile.php';
+        $sql = sprintf(
+            '%s %s',
+            $this->getCommonSQL(),
+            !$this->User->isAdmin || $isProfile
+                ? 'AND calendars.created_by = :user'
+                : '',
+        );
+        $req = $this->Db->prepare($sql);
+        $req->bindValue(':state', State::Normal->value, PDO::PARAM_INT);
+        $req->bindParam(':team', $this->User->team, PDO::PARAM_INT);
+        if (!$this->User->isAdmin || $isProfile) {
+            $req->bindParam(':user', $this->User->userid, PDO::PARAM_INT);
+        }
+        $this->Db->execute($req);
+        $result = $req->fetchAll();
+        $this->postProcessRead($result);
+        return $result;
+    }
+
+    public function readOne(): array
+    {
+        if ($this->id === null) {
+            throw new IllegalActionException('No id was set!');
+        }
+
+        $sql = sprintf(
+            '%s AND id = :id %s',
+            $this->getCommonSQL(),
+            !$this->User->isAdmin
+                ? 'AND calendars.created_by = :user'
+                : '',
+        );
+        $req = $this->Db->prepare($sql);
+        $req->bindValue(':state', State::Normal->value, PDO::PARAM_INT);
+        $req->bindParam(':team', $this->User->team, PDO::PARAM_INT);
+        $req->bindParam(':id', $this->id, PDO::PARAM_INT);
+        if (!$this->User->isAdmin) {
+            $req->bindParam(':user', $this->User->userid, PDO::PARAM_INT);
+        }
+        $this->Db->execute($req);
+        $result = $req->fetchAll();
+        $this->postProcessRead($result);
+        return $result[0];
+    }
+
+    public function destroy(): bool
+    {
+        $this->canOrExplode();
+        $sql = 'UPDATE calendars
+                  SET state = :state
+                  WHERE id = :id';
+        $req = $this->Db->prepare($sql);
+        $req->bindValue(':state', State::Deleted->value, PDO::PARAM_INT);
+        $req->bindParam(':id', $this->id, PDO::PARAM_INT);
+        return $this->Db->execute($req);
+    }
+
+    /**
+     * Generate a random alphanumeric string
+     */
+    public static function randomAlphaNumericString(int $length): string
```

Comment:
I suggest using a uuidv4 instead for tokens: https://github.com/elabftw/elabftw/blob/3000ac3062ea6a9c1c89bc54f90a0cd33f471ee2/src/classes/Tools.php#L33-L42

Reply:
Sure, it is better to use the already existing function.